// 电子鼻 gRPC 服务定义
// 提供控制接口和数据流服务

syntax = "proto3";

package enose.service;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "enose_data.proto";
import "enose_config.proto";

option go_package = "enose/service";

// ============================================================
// 控制服务 (ControlService)
// 用于接收外部控制命令
// ============================================================
service ControlService {
  // 获取系统状态
  rpc GetStatus(google.protobuf.Empty) returns (SystemStatus);
  
  // 切换系统状态
  rpc SetSystemState(SetSystemStateRequest) returns (SetSystemStateResponse);
  
  // 手动控制单个外设
  rpc ManualControl(ManualControlRequest) returns (ManualControlResponse);
  
  // 启动泵 (步进泵需要指定参数)
  rpc RunPump(RunPumpRequest) returns (RunPumpResponse);
  
  // 停止所有泵
  rpc StopAllPumps(google.protobuf.Empty) returns (StopAllPumpsResponse);
  
  // 开始进样 (设置阀门 + 启动指定蠕动泵)
  rpc StartInjection(StartInjectionRequest) returns (StartInjectionResponse);
  
  // 停止进样
  rpc StopInjection(google.protobuf.Empty) returns (StopInjectionResponse);
  
  // 紧急停止 (发送 M112 急停命令，会触发 Klipper shutdown，需要 FIRMWARE_RESTART 恢复)
  rpc EmergencyStop(google.protobuf.Empty) returns (EmergencyStopResponse);
  
  // 重启固件 (急停后恢复)
  rpc FirmwareRestart(google.protobuf.Empty) returns (FirmwareRestartResponse);
  
  // 订阅系统事件流
  rpc SubscribeEvents(google.protobuf.Empty) returns (stream enose.data.Event);
  
  // 订阅外设状态更新
  rpc SubscribePeripheralStatus(google.protobuf.Empty) returns (stream PeripheralStatus);
}

// ============================================================
// 数据服务 (DataService)
// 用于传感器数据流
// ============================================================
service DataService {
  // 订阅传感器数据流
  rpc SubscribeSensorData(google.protobuf.Empty) returns (stream enose.data.SensorFrame);
  
  // 订阅分析结果流
  rpc SubscribeAnalysisResults(google.protobuf.Empty) returns (stream enose.data.AnalysisResult);
}

// ============================================================
// 传感器控制服务 (SensorService)
// 用于控制 BME688 传感器板
// ============================================================
service SensorService {
  // 发送传感器命令 (sync, init, start, stop, status, reset, config)
  rpc SendCommand(SensorCommandRequest) returns (SensorCommandResponse);
  
  // 订阅传感器数据流 (实时推送)
  rpc SubscribeSensorReadings(google.protobuf.Empty) returns (stream SensorReading);
  
  // 获取传感器板状态
  rpc GetSensorStatus(google.protobuf.Empty) returns (SensorBoardStatus);
  
  // 配置加热器
  rpc ConfigureHeater(HeaterConfigRequest) returns (HeaterConfigResponse);
}

// ============================================================
// 请求/响应消息定义
// ============================================================

// 系统状态枚举
enum SystemStateEnum {
  SYSTEM_STATE_UNSPECIFIED = 0;
  INITIAL = 1;        // 开机初始状态
  DRAIN = 2;          // 排废状态
  CLEAN = 3;          // 清洗状态: 出气阀开, 气体三通阀指向大气, 清洗泵开, 夹管阀液路, 排废阀关
  SAMPLE = 4;         // 采样状态: 排废关, 夹管阀气路, 出气阀开, 三通阀指向气室, 气泵开
  INJECT = 5;         // 进样状态: 阀门同CLEAN, 使用蠕动泵进样
}

// 系统状态响应
message SystemStatus {
  // 当前系统状态
  SystemStateEnum current_state = 1;
  
  // 所有外设的当前状态
  PeripheralStatus peripheral_status = 2;
  
  // 系统运行时间 (秒)
  uint64 uptime_seconds = 3;
  
  // 连接状态
  bool moonraker_connected = 4;
  bool sensor_connected = 5;
  bool firmware_ready = 7;  // Klipper 固件是否就绪 (急停后为 false)
  
  // 最后更新时间
  google.protobuf.Timestamp last_updated = 6;
}

// 外设状态
message PeripheralStatus {
  // 阀门状态
  float valve_waste = 1;      // 0: 关闭, 1: 开启
  float valve_pinch = 2;      // 0: 气路, 1: 液路
  float valve_air = 3;        // 0: 排气, 1: 气室
  float valve_outlet = 4;     // 0: 开启, 1: 关闭 (反向逻辑)
  
  // 泵状态
  float air_pump_pwm = 5;     // 0.0 - 1.0
  float cleaning_pump = 6;    // 0.0 - 1.0
  PumpRunState pump_2 = 7;    // 样品泵 0
  PumpRunState pump_3 = 8;    // 样品泵 1
  PumpRunState pump_4 = 9;    // 样品泵 2
  PumpRunState pump_5 = 10;   // 样品泵 3
  
  // 加热器状态
  float heater_chamber = 11;  // 0.0 - 1.0
  
  // 传感器读数 (只读)
  optional float sensor_chamber_temp = 20;  // 气室温度 (°C)
  optional float scale_weight = 21;         // 称重值 (g)
  
  enum PumpRunState {
    PUMP_RUN_STATE_UNSPECIFIED = 0;
    STOPPED = 1;
    RUNNING = 2;
  }
}

// 设置系统状态请求
message SetSystemStateRequest {
  SystemStateEnum target_state = 1;
}

message SetSystemStateResponse {
  bool success = 1;
  string message = 2;
  SystemStateEnum new_state = 3;
}

// 手动控制请求
message ManualControlRequest {
  // 外设名称 (valve_waste, valve_pinch, air_pump_pwm, etc.)
  string peripheral_name = 1;
  
  // 目标值 (对于开关类型: 0/1, 对于PWM: 0.0-1.0)
  float value = 2;
}

message ManualControlResponse {
  bool success = 1;
  string message = 2;
}

// 进样请求
message StartInjectionRequest {
  // 每个蠕动泵的进样量 (单位待标定，目前为步进电机移动距离 mm)
  float pump_2_volume = 1;  // 蠕动泵0
  float pump_3_volume = 2;  // 蠕动泵1
  float pump_4_volume = 3;  // 蠕动泵2
  float pump_5_volume = 4;  // 蠕动泵3
  
  // 可选参数
  optional float speed = 5;  // 进样速度 (mm/s), 默认 10
  optional float accel = 6;  // 加速度 (mm/s²), 默认 100
}

message StartInjectionResponse {
  bool success = 1;
  string message = 2;
}

message StopInjectionResponse {
  bool success = 1;
  string message = 2;
}

// 紧急停止响应
message EmergencyStopResponse {
  bool success = 1;
  string message = 2;
}

// 固件重启响应
message FirmwareRestartResponse {
  bool success = 1;
  string message = 2;
}

// 运行泵请求
message RunPumpRequest {
  // 泵名称 (pump_2, pump_3, pump_4, pump_5, cleaning_pump)
  string pump_name = 1;
  
  // 速度 (步进泵: mm/s, DC泵: 0.0-1.0 PWM)
  float speed = 2;
  
  // 距离 (仅步进泵需要, 单位: mm)
  optional float distance = 3;
  
  // 加速度 (仅步进泵需要, 单位: mm/s²)
  optional float accel = 4;
}

message RunPumpResponse {
  bool success = 1;
  string message = 2;
}

message StopAllPumpsResponse {
  bool success = 1;
  string message = 2;
}

// ============================================================
// 传感器服务消息定义
// ============================================================

// 传感器命令请求
message SensorCommandRequest {
  // 命令类型: sync, init, start, stop, status, reset
  string command = 1;
  
  // 可选参数 (JSON 格式字符串)
  optional string params_json = 2;
}

message SensorCommandResponse {
  bool success = 1;
  string message = 2;
  
  // 响应数据 (JSON 格式字符串)
  optional string data_json = 3;
}

// 传感器板状态
message SensorBoardStatus {
  bool connected = 1;
  bool running = 2;           // 是否正在采集
  uint32 sensor_count = 3;    // 传感器数量
  string firmware_version = 4;
  string port = 5;
}

// 单个传感器读数 (实时数据流)
message SensorReading {
  uint64 tick_ms = 1;         // 设备时间戳 (毫秒)
  uint32 sensor_idx = 2;      // 传感器索引 (0-7)
  uint32 sensor_id = 3;       // 传感器 ID
  double value = 4;           // 主读数 (电阻/电压)
  string sensor_type = 5;     // mox_d, mox_a, pid
  
  optional double temperature = 6;
  optional double humidity = 7;
  optional double pressure = 8;
  
  uint32 heater_step = 9;     // 加热器步进索引
  uint32 adc_channel = 10;
}

// 加热器配置请求
message HeaterConfigRequest {
  repeated uint32 temps = 1;  // 温度列表 (°C)
  repeated uint32 durs = 2;   // 持续时间列表 (×140ms)
  
  // 目标传感器 (空=所有)
  repeated uint32 sensors = 3;
}

message HeaterConfigResponse {
  bool success = 1;
  string message = 2;
}

// 加热器预设
message HeaterPreset {
  string name = 1;
  string description = 2;
  repeated uint32 temps = 3;
  repeated uint32 durs = 4;
}
